docs/rfc/rfc-101-p2p-bad-peers-checktx.md

```diff
@@ -0,0 +1,436 @@
+ Otherwise, we set `lastFailure` to `now` and set `numFailures`  to 1. 
+ Once the value for `numFailures` for a peer reaches `maxAllowedFailures`, the peer is disconnected from and banned. 
+
+The reason for this logic is as follows: We deem it acceptable if every now and then a peer sends us an invalid transaction.
+ But if this happens very frequently, then this behaviour can be considered as spamming and we want to disconnect from the peer. 
+  
+ **Discussion**
+
+ The problem with supporting this scenario is the definition of the above mentioned parameters. It is very hard to estimate, at the CometBFT level, what these parameters should be. A possible solution is 
+ to allow the application to set these parameters. What is unclear, how will the application know that these parameters are not well set if, due to a bug or network problems, transactions start to fail? 
+The network could end up with all nodes banning everyone. How would an application developer know to debug this, what to look for?
+
+A possible solution is to ban peers temporarily. In addition to the question on how long is temporarily, setting specific time limits for banning on a peer basis
+is currently not supported by the p2p layer.
+
+*Banning a peer in case of duplicate transactions*
+
+Currently, a peer can send the same valid (or invalid) transaction [multiple times](https://github.com/tendermint/tendermint/blob/ff0f98892f24aac11e46aeff2b6d2c0ad816701a/mempool/v0/clist_mempool.go#L247). Peers do not 
+gossip transactions to peers that have sent them that same transaction. But there is no check on whether 
+a node has already sent the same transaction to this peer before. There is also no check whether the transaction
+that is being gossiped is currently valid or not (assuming that invalid transactions could become valid).
+The transaction broadcast logic simply loops through the mempool and tries to send the transactions currently in the pool. 
+
+If we want to ban peers based on duplicate transactions, we should either add additional checks for the cases above, or 
+not ban peers for this behaviour at the moment. It would be useful to gather metrics on how often a peer gossips the same 
+transaction and whether this is cause of significant traffic. 
+
+
+#### **Banning for sending *never-valid* transactions**
+
+If a transaction fails since it could never have been valid, `CheckTx` returns a `ResponseCheckTx.code` 
+value greater than 1. In this case, the peer should be disconnected from and banned immediately without keeping count on how often 
+this has happened. 
+
+The question is whether this transaction should be kept track of in the cache? We can still store it in 
+the cache so that we don't run `CheckTx` on it again, but if this peer is immediately banned, maybe there is no need
+to store its information.
```

Comment:
My impression is it doesn't need to be in the cache, since anyone who sends it has a notable punishment, and must keep establishing new peer connections to DOS.

If they were doing a significant effort to establish new peer connections, I believe that with ease, they could create new malicious payloads with costly compute that get banned (rather than reuse the same payload)